testdata/dnn/download_models.py

```diff
@@ -5,215 +5,235 @@
+                    # print('    progress ', end='')
+                    sys.stdout.flush()
+                    while True:
+                        buf = r.read(self.BUFSIZE)
+                        if not buf:
+                            break
+                        f.write(buf)
+                        # print('>', end='')
+                        sys.stdout.flush()
+            # print('')
+            return True
+        except Exception as e:
+            print('  extract failed: {}'.format(e))
+            return False
+
+
+class Processor:
+    def __init__(self, **kwargs):
+        self.reference = kwargs.pop('reference', None)
+        self.verifier = BuiltinVerifier()
+        self.downloader = BuiltinDownloader()
+        self.extractor = BuiltinExtractor()
+
+    def prepare_folder(self, filename):
+        filename.parent.mkdir(parents=True, exist_ok=True)
+
+    def download(self, url, filename):
+        return self.downloader.download(url, filename)
+
+    def verify(self, mdl):
+        return self.verifier.verify(mdl.filename, mdl.sha)
 
-    def ref_copy(self, ref):
-        if not ref:
+    def extract(self, arch, mdl):
+        return self.extractor.extract(arch, mdl.member, mdl.filename)
+
+    def ref_copy(self, mdl):
+        if not self.reference:
+            return False
+        candidate = self.reference / mdl.filename
+        if not candidate.is_file():
             return False
+        print('  ref {} -> {}'.format(candidate, mdl.filename))
+        try:
+            if candidate.absolute() != mdl.filename.absolute():
+                self.prepare_folder(mdl.filename)
+                shutil.copy(candidate, mdl.filename)
+            if self.verify(mdl):
+                return True
+            else:
+                print('  ref - hash mismatch, removing')
+                mdl.filename.unlink()
+                return False
+        except Exception as e:
+            print('  ref failed: {}'.format(e))
+
+    def cleanup(self, filename):
+        print("  cleanup - {}".format(filename))
         try:
-            candidate = ref / self.filename
-            if candidate.is_file():
-                print('  ref {} -> {}'.format(candidate, self.filename))
-                if candidate.absolute() == self.filename.absolute():
-                    print('  ref - skip copying the same file')
-                else:
-                    self.prepare_file()
-                    shutil.copy(candidate, self.filename)
-                if self.verify():
-                    return True
-                else:
-                    print('  ref - hash mismatch, removing')
-                    self.filename.unlink()
+            filename.unlink()
         except Exception as e:
-            print('  ref {}'.format(e))
+            print("  cleanup failed: {}".format(e))
 
-    def get_sub(self, ref, arch):
-        print('** {}'.format(self.filename))
+    def handle_bad_download(self, filename):
```

Comment:
This function has been reduced, I hope the logic is similar to the original. Maybe some cornercases are not handled well, but in general I think it is OK.